src/Magnum/BulletIntegration/MotionState.cpp

```diff
@@ -26,27 +26,50 @@
 
 #include "MotionState.h"
 
+#include <Magnum/Math/Angle.h>
 #include <Magnum/Math/Matrix4.h>
 
 #include "Magnum/BulletIntegration/Integration.h"
 
+
 namespace Magnum { namespace BulletIntegration {
 
+using namespace Math::Literals;
+
 void MotionState::getWorldTransform(btTransform& worldTrans) const {
-    const Matrix4 transformation = object().transformationMatrix();
-    worldTrans.setOrigin(btVector3(transformation.translation()));
-    worldTrans.setBasis(btMatrix3x3(transformation.rotationScaling()));
+    /* Bullet transforms are always world space */
+    worldTrans = btTransform{object().absoluteTransformationMatrix()};
 }
 
 void MotionState::setWorldTransform(const btTransform& worldTrans) {
-    btVector3 bPosition = worldTrans.getOrigin();
-    btVector3 bAxis = worldTrans.getRotation().getAxis();
-    Rad bRotation(worldTrans.getRotation().getAngle());
-
-    /** @todo Verify that all objects have common parent */
-    transformation.resetTransformation()
-        .rotate(bRotation, Vector3(bAxis).normalized())
-        .translate(Vector3(bPosition));
+    auto parent = object().parent();
+    if(parent) {
+        /* Get local transform Bullet world transform by removing parent transform */
+        const Matrix4 parentTransform = parent->absoluteTransformationMatrix();
+
+        CORRADE_ASSERT(parentTransform.isRigidTransformation(), "MotionState::setWorldTransform(): parent of object with MotionState cannot have scaling", );
+        const Quaternion parentRotation = Quaternion::fromMatrix(parentTransform.rotationScaling());
+
+        const Quaternion rotation = (parentRotation.inverted()*Quaternion{worldTrans.getRotation()})
+                                    .normalized(); /* Avoid floating point drift */
+        const Vector3 translation = Vector3{worldTrans.getOrigin()} - parentTransform.translation();
```

Comment:
Hmm. All the inversions, `sqrt()` when converting the matrix to quaternion and renormalizations are quite costly operations and I'm not sure how this would perform with 100s of objects.

But on the other hand I also don't see any better way how to handle this if you really want to cover the general case where every object can have a different parent but you still want them all in a single physics world.

Things to consider:

- there is transformation caching that could be maybe used for this purpose ([docs](http://mosra.cz/blog/magnum-doc/scenegraph.html#scenegraph-features-caching)) assuming that the parent transformation doesn't change that much (otherwise it would be just more overhead)
- there's also a way to get absolute transformations of a bunch of objects at once (not just one) using [SceneGraph::Object](http://mosra.cz/blog/magnum-doc/classMagnum_1_1SceneGraph_1_1Object.html#a87a8cd6facd28a091523f5b35c6e86a9) and that would also save some calculation time when more objects share the same parent. Does Bullet have some way to batch update transformations of more objects where this could be used? Unfortunately there's no way to get inverse absolute transformations (yet), I would need to add that.